synapse/storage/engines/postgres.py

```diff
@@ -72,33 +81,31 @@ def check_database(self, db_conn, allow_outdated_version: bool = False):
                     "See docs/postgres.md for more information." % (rows[0][0],)
                 )
 
-            txn.execute(
-                "SELECT datcollate, datctype FROM pg_database WHERE datname = current_database()"
-            )
-            collation, ctype = txn.fetchone()
+            collation, ctype = self.get_db_locale(txn)
             if collation != "C":
-                logger.warning(
-                    "Database has incorrect collation of %r. Should be 'C'\n"
-                    "See docs/postgres.md for more information.",
-                    collation,
-                )
+                if not allow_unsafe_locale:
```

Comment:
Would be tempted to flatten this nested if to `if collation != "C" and not allow_unsafe_locale:`.
Actually, scratch that, I'd be tempted to still log a warning when the config flag is enabled, just thinking that it's good to have the log file reveal when a 'tainted configuration' is being used.